addons/social_share/models/snshare_campaign.py

```diff
@@ -0,0 +1,262 @@
+                render_values,
+                raise_on_code=False,
+            ))
+        return bodies
+
+    def _get_images_bytes(self, record=None, with_demo_values=False):
+        bodies = self._get_bodies(record=record, with_demo_values=with_demo_values)
+        if self and bodies:
+            return self.env['ir.actions.report']._run_wkhtmltoimage(
+                bodies,
+                *TEMPLATE_DIMENSIONS
+            )
+        return [b''] * len(self)
+
+    def _get_images_b64(self, record=None, with_demo_values=False):
+        return [
+            base64.b64encode(image_bytes) for image_bytes
+            in self._get_images_bytes(record=record, with_demo_values=with_demo_values)
+        ]
+
+    @api.onchange('template_group_id')
+    def _onchange_template_group_id(self):
+        for campaign in self:
+            campaign.template_id = campaign.template_group_id.template_ids[:1]
+
+    @api.onchange('template_id')
+    def _onchange_template_id(self):
+        # reset color so template default is applied
+        self.element_ids.text_color = False
+
+    @api.onchange('model')
+    def _onchange_model(self):
+        # force trigger so it behaves nicely in the UI
+        self.element_ids._onchange_model()
+
+    def action_open_url_share(self):
+        """Open url dialog."""
+        self.ensure_one()
+        return {
+            'type': 'ir.actions.act_window',
+            'name': _('Share Link'),
+            'res_model': 'snshare.url.share',
+            'views': [[False, 'form']],
+            'context': {
+                'default_campaign_id': self.id,
+                'dialog_size': 'medium',
+            },
+            'target': 'new',
+        }
+
+    def action_show_clicked_urls(self):
+        self.ensure_one()
+        return self.env["ir.actions.actions"]._for_xml_id("social_share.action_snshare_url") | {
+            'context': {'search_default_filter_visited': True},
+            'domain': [('campaign_id', '=', self.id)],
+        }
+
+    def action_show_shared_urls(self):
+        self.ensure_one()
+        return self.env["ir.actions.actions"]._for_xml_id("social_share.action_snshare_url") | {
+            'context': {'search_default_filter_shared': True},
+            'domain': [('campaign_id', '=', self.id)],
+        }
+
+    def action_share_multi(self):
+        self.ensure_one()
+        return {
+            'type': 'ir.actions.act_window',
+            'name': _('Send Share URLS'),
+            'res_model': 'snshare.url.share.multi',
+            'context': {'default_share_campaign_id': self.id, 'default_subject': self.name},
+            'views': [[False, 'form']],
+            'target': 'new',
+        }
+
+    def _get_redirect_url(self, uuid):
+        self.ensure_one()
+        base = self.get_base_url()
+        return f'{base}/snshare/redirect/{self.id}' + (f'/{uuid}' if uuid else '')
+
+    def _get_campaign_url(self, uuid):
```

Comment:
Isn't it the card url, as we give an uuid ? Shouldn't we have this method on the "url" (card ?) model ? Seems to imply several redirections, notably in the share wizard (e.g. could do wizard.card_id.url instead of calling _get_campaign_url, quickly looking at code).